program/tests/tests/add_remove_validator.rs

```diff
@@ -43,11 +48,57 @@ async fn test_remove_validator_without_unclaimed_credits() {
     let vote_account = solido.validators.entries[0].pubkey;
     assert_eq!(solido.validators.entries[0].entry.fee_credit, StLamports(0));
 
-    context
-        .try_remove_validator(vote_account)
-        .await
-        .expect("Failed to remove validator.");
+    let result = context.try_remove_validator(vote_account).await;
 
-    let solido = context.get_solido().await;
-    assert_eq!(solido.validators.len(), 0);
+    assert_solido_error!(result, LidoError::ValidatorIsStillActive);
+}
+
+#[tokio::test]
+async fn test_removing_validator_with_stake_accounts_should_fail() {
+    let mut context = Context::new_with_maintainer().await;
+    let validator = context.add_validator().await;
+
+    // Sanity check before we start: the validator should have zero balance in zero stake accounts.
+    let solido_before = context.get_solido().await;
+    let validator_before: &Validator = &solido_before.validators.entries[0].entry;
+    assert_eq!(validator_before.stake_accounts_balance, Lamports(0));
+    assert_eq!(validator_before.stake_seeds.stake_accounts_seed_begin, 0);
+    assert_eq!(validator_before.stake_seeds.stake_accounts_seed_end, 0);
+
+    // Now we make a deposit, and then delegate part of it.
+    context.deposit(TEST_DEPOSIT_AMOUNT).await;
+
+    let stake_account = context
+        .stake_deposit(
+            validator.vote_account,
+            StakeDeposit::Append,
+            TEST_STAKE_DEPOSIT_AMOUNT,
+        )
+        .await;
+
+    // The amount that we staked, should now be in the stake account.
+    assert_eq!(
+        context.get_sol_balance(stake_account).await,
+        TEST_STAKE_DEPOSIT_AMOUNT
+    );
+
+    // We should also have recorded in the Solido state that this validator now
+    // has balance in a stake account.
+    let solido_after = context.get_solido().await;
+
+    let validator_after = &solido_after.validators.entries[0].entry;
+    assert_eq!(
+        validator_after.stake_accounts_balance,
+        TEST_STAKE_DEPOSIT_AMOUNT
+    );
+
+    // This was also the first deposit, so that should have created one stake account.
+    assert_eq!(validator_after.stake_seeds.stake_accounts_seed_begin, 0);
+    assert_eq!(validator_after.stake_seeds.stake_accounts_seed_end, 1);
+
+    let result = context.try_remove_validator(validator.vote_account).await;
+
+    // The validator should not be able to be removed if it is still active
+    //  (i.e. the active flag is set toe true OR it has stake accounts)
+    assert_solido_error!(result, LidoError::ValidatorIsStillActive);
```

Comment:
If removal fails on the fact that the validator is still active, then all of this business above of giving it stake was not relevant. We should also deactivate the validator (which does not deactivate the stake), and then we can test that it’s really the stake that prevents removing. The function to deactivate is here: https://github.com/chorusone/solido/blob/295a263525fb20c0e17c01bae994c76dd74c4d12/program/tests/context.rs#L707